### PR TITLE
simple-zmq: Throw (zmq-error) at error.

### PR DESCRIPTION
**Pull request by RoubyPA**
Hi Jerry,

Sorry for this huge commit...
I have rewrite the options list and error management.

Now we can get `zmq_error` exception with catch:
``` scheme
(catch 'zmq_error
   (lambda ()
     ...)  ;Your code with simple-zmq procedure
   (lambda stuff
     ...)) ;Error handler with stuff list like '(zmq_error errno strerror)
```

It's possible have to made change on project who use `simple-zmq`.
For example change socket type from `'ZMQ_PUB` to `ZMQ_PUB` at socket creation.

Please, tell me if you want I made modification.

### CODE REVIEW

**Comment by jerry40** [comment]
Hi! Thanks for your work! I'd like to test new version with guile jupyter kernel, I can't do this at the moment, I'll be back in few days!

**Comment by jerry40** [comment]
I think the changes are very good. I used 'ZMQ-PUB notation instead of ZMQ-PUB variable because I thought that having hundreds of variables is not a good idea. But actually it can be a good idea in spite of  autocomplete features. An ideal decision would be to create 'new' datatype to distinguish socket types from socket options, for example, but I don't know how to do this in quile and I don't think it is worth spending time.
Guile jupyter kernel will require easy changes and I'm sure not many programmers use simple-zmq besides the jupyter kernel. So I guess I'll add CHANGELOG.md and change code examples in README. And I have not tested it on my side yet. Will do this in next days. Thank you very much!

**Comment by jerry40** [comment]
Hi! I get this error:

ERROR: zmq_ctx_new: import-func failed: (wrong-number-of-args #f "Wrong number of arguments to ~A" (#<procedure pointer->procedure (_ _ _)>) #f)

Looks like my version of pointer->procedure does not accept the optional parameter.

guile (GNU Guile) 2.0.11
Packaged by Debian (2.0.11-deb+1-10)

Is this errno parameter a new feature?

**Comment by RoubyPA** [comment]
Hi,

Thanks for review !

Yeah, add a `CHANGELOG.md` could be useful.

By the way, I have not tested all `ZMQ` feature, but I think it would be useful to add build system and test suite for `simple-zmq`.

---

This feature are appear at [Guile 2.0.12](https://lists.gnu.org/archive/html/guile-user/2016-07/msg00036.html).

```
Changes in 2.0.12 (since 2.0.11):

* Notable changes

** FFI: Add support for functions that set 'errno'

When accessing POSIX functions from a system's libc via Guile's dynamic
FFI, you commonly want to access the 'errno' variable to be able to
produce useful diagnostic messages.

This is now possible using 'pointer->procedure' or
'scm_pointer_to_procedure_with_errno'.  See "Dynamic FFI" in the manual.
```

It is possible to update your guile ([Debian Guile](https://packages.debian.org/stable/guile-2.0)) ? 

I thinks is very tricky to get `errno` without this feature. :thinking:

**Comment by jerry40** [comment]
Yes, let me upgrade my system ant continue testing then. Hopefully, no one uses guile < 2.0.12. Perhaps we can keep the previous version for those who uses old guile... I don't know how to do this correctly though.

**Comment by RoubyPA** [comment]
I think is possible to make like [shepherd](https://git.savannah.gnu.org/cgit/shepherd.git)  with `syscall->procedure` ([system.scm.in](https://git.savannah.gnu.org/cgit/shepherd.git/tree/modules/shepherd/system.scm.in)).

**Comment by jerry40** [comment]
Hi! I was able to refresh my environment entirely, so hopefully I'll test your comit soon.

File: src/simple-zmq.scm [hunk 1]
```diff
@@ -66,119 +251,139 @@
...
+;;
+;; Error.
+;;
+
+(define (zmq-get-strerror errno)
+  (pointer->string (zmq_strerror errno)))
+
+(define (zmq-get-error errno)
+  (let ((strerror (zmq-get-strerror errno)))
+    (display "ZMQ: error\n")
```
Comment (RoubyPA) [inline src/simple-zmq.scm:263]:
```scheme
(display "ZMQ: error\n")
```
I thinks we could delete this line before merge.

**Comment by jerry40** [comment]
Testing your version and what I noticed - first execution of the jupyter guile kernel takes 1 minute! Good news - it is not about your changes since the delay with the previous version is the same. Will try to find the problem bit later.
Updating the system is always painful :)

**Comment by jerry40** [comment]
Ok, I've tested it and it seems to be ok! Thanks!

**Pull request merged**